### PR TITLE
FIX-670: manual fixes for specifically mentioned ABI

### PR DESCRIPTION
**Pull request by DenisYaroshevskiy**
These fixes are not as dumb, but I refactor a bit and also add support for avx512 where it's easy.

in max/ min for avx went to if_else for 4 byte integers since there is a blend.

### CODE REVIEW

File: include/eve/function/add.hpp [hunk 1]
```diff
@@ -91,3 +91,7 @@ namespace eve
 }
 
 #include <eve/module/real/core/function/regular/generic/add.hpp>
+
+#if defined(EVE_INCLUDE_X86_HEADER)
+#  include <eve/module/real/core/function/regular/simd/x86/floor.hpp>
```
Comment (jfalcou) [inline include/eve/function/add.hpp:96]: why adding floor to add ?

File: include/eve/module/real/core/function/regular/simd/x86/average.hpp [hunk 2]
```diff
@@ -16,32 +16,24 @@
...
-  EVE_FORCEINLINE wide<T, N, x86_256_> average_(EVE_SUPPORTS(avx_)
-                                           , wide<T, N, x86_256_> const &a
-                                           , wide<T, N, x86_256_> const &b) noexcept
-  {
-    if constexpr(current_api >= avx2)
+    if constexpr ( sizeof(T) == 2 )
     {
-             if constexpr(sizeof(T) == 1) return _mm256_avg_epu8(a, b);
-        else if constexpr(sizeof(T) == 2) return _mm256_avg_epu16(a, b);
-        else                              return (a & b) + ((a ^ b) >> 1);
```
Comment (jfalcou) [inline include/eve/module/real/core/function/regular/simd/x86/average.hpp:42]: where does this one went ?

File: include/eve/module/real/core/function/regular/simd/x86/ceil.hpp [hunk 3]
```diff
@@ -20,21 +20,20 @@ namespace eve::detail
   //-----------------------------------------------------------------------------------------------
   // 128 bits implementation
   template<floating_real_scalar_value T, typename N, typename ABI>
-  EVE_FORCEINLINE wide<T, N, x86_128_> ceil_(EVE_SUPPORTS(sse4_1_)
-                                        , wide<T, N, x86_128_> const &a0) noexcept
```
Comment (jfalcou) [inline include/eve/module/real/core/function/regular/simd/x86/ceil.hpp:24]: will we change those to if constexpr later ?

File: include/eve/module/real/core/function/saturated/simd/x86/add.hpp [hunk 4]
```diff
@@ -14,103 +14,37 @@
...
+    else if constexpr ( std::same_as<abi_t<T, N>, x86_256_> && current_api == avx ) return add_(EVE_RETARGET(cpu_), st, v0, v1);
+    else if constexpr ( std::signed_integral<T> && sizeof(T) == 1 )
     {
-      return add(v0, v1);
+           if constexpr ( std::same_as<abi_t<T, N>, x86_512_> ) return _mm512_adds_epi8(v0, v1);
+      else if constexpr ( std::same_as<abi_t<T, N>, x86_256_> ) return _mm256_adds_epi8(v0, v1);
+      else if constexpr ( std::same_as<abi_t<T, N>, x86_128_> ) return _mm_adds_epi8   (v0, v1);
     }
-    else
+    else if constexpr ( std::signed_integral<T> && sizeof(T) == 2 )
```
Comment (jfalcou) [inline include/eve/module/real/core/function/saturated/simd/x86/add.hpp:31]: not sure we'll have float of size 2 on x86 so th eis_integral is maybe useless

**Review by jfalcou — CHANGES_REQUESTED** [review]

File: include/eve/detail/function/simd/x86/combine.hpp [hunk 5]
```diff
@@ -7,155 +7,92 @@
...
-      else  if constexpr( std::integral<T> )
-      {
-        return _mm_castps_si128(
-            _mm_shuffle_ps(_mm_castsi128_ps(l.storage()), _mm_castsi128_ps(h.storage()), 0x44));
-      }
+      else                                          return _mm_castps_si128(_mm_shuffle_ps(_mm_castsi128_ps(l),
+                                                                            _mm_castsi128_ps(h),
+                                                                            0x44));
     }
-    //==============================================================================================
```
[the remark this answers is not in the document]
Comment (DenisYaroshevskiy) [reply include/eve/detail/function/simd/x86/combine.hpp:44]: I don't see why in the world they are useful

File: include/eve/module/real/core/function/regular/simd/x86/ceil.hpp (hunk 3 above)
Comment (DenisYaroshevskiy) [reply include/eve/module/real/core/function/regular/simd/x86/ceil.hpp:24]: I just didn't find the funcitons. Will fix.

File: include/eve/module/real/core/function/saturated/simd/x86/add.hpp (hunk 4 above)
Comment (DenisYaroshevskiy) [reply include/eve/module/real/core/function/saturated/simd/x86/add.hpp:31]: This is about `signed` and `unsigned` not about floats

File: include/eve/module/real/core/function/regular/simd/x86/ceil.hpp (hunk 3 above)
Comment (DenisYaroshevskiy) [reply include/eve/module/real/core/function/regular/simd/x86/ceil.hpp:24]: Done

File: include/eve/module/real/core/function/regular/simd/x86/frac.hpp [hunk 6]
```diff
@@ -35,12 +35,12 @@ namespace eve::detail
   // -----------------------------------------------------------------------------------------------
   // 256 bits implementation for xop
   template<floating_real_scalar_value T, typename N>
-  EVE_FORCEINLINE auto frac_(EVE_SUPPORTS(avx_),
-                             wide<T, N, x86_256_> const &a0) noexcept
+  EVE_FORCEINLINE wide<T, N> frac_(EVE_SUPPORTS(avx_), wide<T, N> const &a0) noexcept
```
Comment (jfalcou) [inline include/eve/module/real/core/function/regular/simd/x86/frac.hpp:38]: why the change from `auto`?

Comment (DenisYaroshevskiy) [reply include/eve/module/real/core/function/regular/simd/x86/frac.hpp:38]: Because `_mm256_frcz_pd` does not return `wide`

File: include/eve/module/real/core/function/regular/simd/x86/average.hpp (hunk 2 above)
Comment (DenisYaroshevskiy) [reply include/eve/module/real/core/function/regular/simd/x86/average.hpp:42]: CPU should  do this https://github.com/jfalcou/eve/blob/1e865bfb459ea3dfb4a6c9b10944c55ae761a841/include/eve/module/real/core/function/regular/generic/average.hpp#L42

**Review by jfalcou — APPROVED** [review]
